### PR TITLE
dev/core#87 PayPal Pro missing parameters (add standard getters)

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
state_province, country and ip_address are not always set properly for paypal pro due to inconsistency with parameters being passed.
Ref: https://lab.civicrm.org/dev/core/issues/87

Before
----------------------------------------
state_province, country and ip_address may not be passed to payment processor.

After
----------------------------------------
This applies to ALL payment processors:
* If state_province and country were passed from the payment form in one way or another, they will be presented in a consistent format to the payment processor.
* Add client ip_address to the list of parameters so payment processors don't have to retrieve this manually if they require it (paypal pro wants it but was not retrieving it).

Technical Details
----------------------------------------
This PR adds three generic "getters" to CRM_Core_Payment per @eileenmcnaughton suggestion.  For compatibility, no parameters are removed but additional parameters may be set depending on what is passed in.  For example if only country_id is defined, then country will be set as well.

Comments
----------------------------------------
This fixes a real issue identified with the paypal pro processor in core, but the inconsistency of parameters passed affects other processors and means each processor has to do extra coding to make sure that the parameters are in the right format when they are passed in.
See for example https://civicrm.stackexchange.com/questions/24316/paypal-pro-billing-address-state-province-country-not-being-passed

### CODE REVIEW

File: CRM/Core/Payment.php [hunk 1]
```diff
@@ -1181,6 +1227,11 @@ public function doPayment(&$params, $component = 'contribute') {
       return $result;
     }
 
+    // Set some "Standard" parameters for payment
+    $this->getIpAddress($params);
```
Comment (eileenmcnaughton) [inline CRM/Core/Payment.php]: I feel like calling this functions should be done in the payment processor classes - rather than here

Comment (eileenmcnaughton) [reply CRM/Core/Payment.php]: (I'm OK with adding the functions per above)

Comment (seamuslee001) [reply CRM/Core/Payment.php]: @eileenmcnaughton why wouldn't this happen here, just thinking if these fields applied equally to all payment processors why wouldn't as core have it all in one place?

Comment (eileenmcnaughton) [reply CRM/Core/Payment.php]: @seamuslee001 well 
a) there is an existing pattern of processors using the get function when they want to get certain details (getAmount for example)
b) we actually encourage processors to entirely override doPayment these days
c) we are mixing get & set in the same action - ie. we have $params but we are doing a set within it calling get

Note that the issue here is that core CiviCRM actually ensures these things are set before calling doPayment (I don't know what gaps there are) as does webform - but the goal is to move that out of webform into these classes. Perhaps at some point also out of the core forms but really what we need is a much clearer 'contract' as to what is required when calling doPayment

**Comment by eileenmcnaughton** [comment]
Actually on re-reading - I think our getters should JUST get - I realise they are actually setting. Ideally we would have a setParameters function & then $this->parameters would be used rather than passing $params around - but we can't do that in the construct fn which would have made sense :-(

**Comment by eileenmcnaughton** [comment]
@mattwire so here is how I would think this might look  - it would take some time before the payment processors adopt it & if your short-term goal is to get webform to be formatting the params as the processors expect them it should call the same sorts of functions that core does in the short-tesm

    list($hasBillingField, $addressParams) = CRM_Contribute_BAO_Contribution::getPaymentProcessorReadyAddressParams($this->_params, $this->_bltID);
    $fields = $this->formatParamsForPaymentProcessor($fields);

(note I'll close this until you have updated for re-review. While discussing we can keep it closed)

```
  /**
-   * Get the currency for the transaction.
-   *
-   * Handle any inconsistency about how it is passed in here.
+ * Get the currency for the transaction.
+ *
+ * Handle any inconsistency about how it is passed in here.
+ *
+ * @param $params
+ *
+ * @return string
+ */
+  protected function getCurrency($params) {
+    return CRM_Utils_Array::value('currencyID', $params, CRM_Utils_Array::value('currency', $params));
+  }
+
+  /**
+   * Get the IP address of the remote user.
    *
    * @param $params
    *
    * @return string
    */
-  protected function getCurrency($params) {
-    return CRM_Utils_Array::value('currencyID', $params, CRM_Utils_Array::value('currency', $params));
+  protected function getIpAddress($params) {
+    return CRM_Utils_Array::value('ip_address', $params, RM_Utils_System::ipAddress());
+  }
+
+  /**
+   * Get the state_province.
+   *
+   * @param $params
+   *
+   * @return array
+   */
+  protected function getStateProvinceName($params) {
+
+  }
+
+  /**
+   * Get the state_province abbreviation.
wmf1411:civicrm emcnaughton$ git diff
diff --git a/CRM/Core/Payment.php b/CRM/Core/Payment.php
index 202e3fe729..c8ad2c9401 100644
--- a/CRM/Core/Payment.php
+++ b/CRM/Core/Payment.php
@@ -986,16 +986,69 @@ abstract class CRM_Core_Payment {
   }
 
   /**
-   * Get the currency for the transaction.
-   *
-   * Handle any inconsistency about how it is passed in here.
+ * Get the currency for the transaction.
+ *
+ * Handle any inconsistency about how it is passed in here.
+ *
+ * @param $params
+ *
+ * @return string
+ */
+  protected function getCurrency($params) {
+    return CRM_Utils_Array::value('currencyID', $params, CRM_Utils_Array::value('currency', $params));
+  }
+
+  /**
+   * Get the IP address of the remote user.
    *
    * @param $params
    *
    * @return string
    */
-  protected function getCurrency($params) {
-    return CRM_Utils_Array::value('currencyID', $params, CRM_Utils_Array::value('currency', $params));
+  protected function getIpAddress($params) {
+    return CRM_Utils_Array::value('ip_address', $params, RM_Utils_System::ipAddress());
+  }
+
+  /**
+   * Get the state_province.
+   *
+   * It looks like BAO_Contribution::getBillingAddressParams ensures that $params['state_province_id']
+   * is set so we can rely on that.
+   *
+
+  /**
+   * Get the country.
+   *
+   * It looks like BAO_Contribution::getBillingAddressParams ensures that $params['country_id']
+   * is set so we can rely on that.
+   *
+   * @param $params
+   *
+   * @return array
+   */
+  protected function getBillingCountryIsoCode($params) {
+    return CRM_Core_PseudoConstant::countryIsoCode($params['country_id']);
   }
 
   /**

```

**Pull request closed without merging**